File: core/moduleloader/src/main/java/com/gamecenter/app/moduleloader/ModuleResourceLoader.java
package com.gamecenter.app.moduleloader;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.res.AssetManager;
import android.content.res.Resources;
import android.util.DisplayMetrics;
import android.util.Log;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java.io.File;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;

/**
 * 模块资源加载器（增强版）。
 * 
 * 负责加载模块 APK 中的资源（布局、图片、字符串等）。
 * 使用 AssetManager.addAssetPath() 加载插件资源。
 * 
 * @author Software Engineer (Alex)
 * @version 2.0
 * @since 2026-05-26
 */
public class ModuleResourceLoader {
    
    private static final String TAG = "ModuleResourceLoader";
    
    /** 资源缓存（模块 ID -> Resources） */
    private final Map<String, Resources> resourcesCache;
    
    /** Context 引用 */
    private final Context context;
    
    /** 主包名（框架 APK 的包名） */
    private String mainPackageName;
    
    /**
     * 构造函数。
     * 
     * @param context Android Context
     */
    public ModuleResourceLoader(@NonNull Context context) {
        this.context = context != null ? context.getApplicationContext() : null;
        this.resourcesCache = new HashMap<>();
        this.mainPackageName = this.context != null ? 
                this.context.getPackageName() : "com.gamecenter.app";
    }
    
    /**
     * 加载模块 APK 的资源。
     * 
     * @param apkFile 模块 APK 文件
     * @return 模块资源的 Resources 对象，失败返回 null
     */
    @Nullable
    public Resources loadResources(@NonNull File apkFile) {
        if (context == null) {
            Log.e(TAG, "Context 为 null");
            return null;
        }
        
        if (apkFile == null || !apkFile.exists()) {
            Log.e(TAG, "APK 文件不存在: " + apkFile);
            return null;
        }
        
        // 检查缓存
        String cacheKey = apkFile.getAbsolutePath();
        Resources cached = resourcesCache.get(cacheKey);
        if (cached != null) {
            Log.d(TAG, "使用缓存的 Resources: " + apkFile.getName());
            return cached;
        }
        
        try {
            // 创建 AssetManager
            AssetManager assetManager = createAssetManager(apkFile.getAbsolutePath());
            if (assetManager == null) {
                return null;
            }
            
            // 获取主包的 Resources 作为配置模板
            Resources mainResources = context.getResources();
            if (mainResources == null) {
                Log.e(TAG, "主 Resources 为 null");
                return null;
            }
            
            // 创建新的 Resources (使用复制的 Configuration 避免污染宿主全局配置)
            android.content.res.Configuration config = new android.content.res.Configuration(mainResources.getConfiguration());
            Resources moduleResources = new Resources(
                    assetManager,
                    mainResources.getDisplayMetrics(),
                    config
            );
            
            // 缓存
            resourcesCache.put(cacheKey, moduleResources);
            
            Log.d(TAG, "资源加载成功: " + apkFile.getName());
            return moduleResources;
            
        } catch (Exception e) {
            Log.e(TAG, "加载资源失败: " + apkFile.getName(), e);
            return null;
        }
    }
    
    /**
     * 创建 AssetManager 并添加 APK 路径。
     * 
     * @param apkPath APK 文件路径
     * @return AssetManager 实例，失败返回 null
     */
    @Nullable
    private AssetManager createAssetManager(@NonNull String apkPath) {
        try {
            AssetManager assetManager = AssetManager.class.newInstance();
            
            // 调用 addAssetPath(String path)
            Method addAssetPath = AssetManager.class.getMethod("addAssetPath", String.class);
            int result = (int) addAssetPath.invoke(assetManager, apkPath);
            
            if (result == 0) {
                Log.e(TAG, "addAssetPath 失败: " + apkPath);
                return null;
            }
            
            Log.d(TAG, "AssetManager 创建成功，已添加路径: " + apkPath);
            return assetManager;
            
        } catch (Exception e) {
            Log.e(TAG, "创建 AssetManager 失败", e);
            return null;
        }
    }
    
    /**
     * 从模块 APK 中加载 Drawable。
     * 
     * @param apkFile 模块 APK 文件
     * @param resId 资源 ID
     * @return Drawable 对象，失败返回 null
     */
    @Nullable
    public android.graphics.drawable.Drawable loadDrawable(@NonNull File apkFile, int resId) {
        Resources resources = loadResources(apkFile);
        if (resources == null) {
            return null;
        }
        
        try {
            return resources.getDrawable(resId, context.getTheme());
        } catch (Exception e) {
            Log.e(TAG, "加载 Drawable 失败: resId=0x" + Integer.toHexString(resId), e);
            return null;
        }
    }
    
    /**
     * 从模块 APK 中加载布局（简化实现，使用模块资源 Context）。
     * 
     * @param apkFile 模块 APK 文件
     * @param layoutId 布局 ID
     * @return View 对象，失败返回 null
     */
    @Nullable
    public android.view.View loadLayout(@NonNull File apkFile, int layoutId) {
        // 调用新的重载方法，parent 传 null
        return loadLayout(apkFile, layoutId, null);
    }
    
    /**
     * 从模块 APK 中加载字符串。
     * 
     * @param apkFile 模块 APK 文件
     * @param resId 字符串资源 ID
     * @return 字符串，失败返回 null
     */
    @Nullable
    public String loadString(@NonNull File apkFile, int resId) {
        Resources resources = loadResources(apkFile);
        if (resources == null) {
            return null;
        }
        
        try {
            return resources.getString(resId);
        } catch (Exception e) {
            Log.e(TAG, "加载字符串失败: resId=0x" + Integer.toHexString(resId), e);
            return null;
        }
    }
    
    /**
     * 获取模块 APK 的包名（简化实现）。
     * 
     * @param apkFile 模块 APK 文件
     * @return 包名，失败返回 null
     */
    @Nullable
    public String getModulePackageName(@NonNull File apkFile) {
        // 简化实现：实际应从 AndroidManifest.xml 解析
        // 约定：模块包名为 com.gamecenter.module.{moduleId}
        Log.w(TAG, "getModulePackageName 为简化实现，需完善");
        return null;
    }
    
    /**
     * 清除指定模块的 Resources 缓存。
     * 
     * @param apkPath APK 文件路径
     */
    public void clearCache(@NonNull String apkPath) {
        resourcesCache.remove(apkPath);
        Log.d(TAG, "已清除资源缓存: " + apkPath);
    }
    
    /**
     * 清除所有 Resources 缓存。
     */
    public void clearAllCache() {
        resourcesCache.clear();
        Log.d(TAG, "已清除所有资源缓存");
    }
    
    /**
     * 获取当前缓存的模块数量。
     * 
     * @return 缓存数量
     */
    public int getCachedModuleCount() {
        return resourcesCache.size();
    }
    
    /**
     * 创建模块资源的 Context（用于正确加载模块布局）。
     * 
     * @param apkFile 模块 APK 文件
     * @param baseContext 基础 Context
     * @return 包含模块资源的 Context，失败返回 baseContext
     */
    @NonNull
    public Context createModuleContext(@NonNull File apkFile, @NonNull Context baseContext) {
        Resources moduleResources = loadResources(apkFile);
        if (moduleResources == null) {
            Log.w(TAG, "无法加载模块资源，使用基础 Context");
            return baseContext;
        }
        
        // 创建 ContextWrapper，重写 getResources() 返回模块资源
        return new ContextWrapper(baseContext) {
            @Override
            public Resources getResources() {
                return moduleResources;
            }
            
            @Override
            public AssetManager getAssets() {
                return moduleResources.getAssets();
            }
            
            @Override
            public Context getApplicationContext() {
                return baseContext.getApplicationContext();
            }
        };
    }
    
    /**
     * 从模块 APK 中加载布局（使用模块资源 Context）。
     * 
     * @param apkFile 模块 APK 文件
     * @param layoutId 布局 ID
     * @param parent 父 View（可选，用于正确解析 LayoutParams）
     * @return View 对象，失败返回 null
     */
    @Nullable
    public android.view.View loadLayout(@NonNull File apkFile, int layoutId, 
                                        @Nullable android.view.ViewGroup parent) {
        Context moduleContext = createModuleContext(apkFile, context);
        
        try {
            android.view.LayoutInflater inflater = android.view.LayoutInflater.from(moduleContext);
            return inflater.inflate(layoutId, parent, false);
        } catch (Exception e) {
            Log.e(TAG, "加载布局失败: layoutId=0x" + Integer.toHexString(layoutId), e);
            return null;
        }
    }
}

File: core/module-host/src/main/kotlin/com/gamecenter/app/core/modulehost/ModuleResourceLoader.kt
package com.gamecenter.app.core.modulehost

import android.content.Context
import android.content.res.AssetManager
import android.content.res.Resources
import android.graphics.drawable.Drawable
import android.util.Log
import java.io.File

/**
 * 外部模块资源加载器（统一真源）。
 *
 * 职责：为动态模块 APK 构造独立 [Resources]，使模块布局与资源能够通过
 * 模块自带的 AssetManager 解析。实现自 `app/modular/ModuleResourceLoader` 收敛而来，
 * 位于加载器所在核心层，宿主侧不再维护副本。
 */
class ModuleResourceLoader(
    private val context: Context
) {

    private val resourceCache = mutableMapOf<String, ModuleResources>()

    companion object {
        /** 一次性探测结果缓存（进程内）。 */
        @Volatile
        private var addAssetPathAvailable: Boolean? = null

        private const val TAG = "ModuleResourceLoader"

        /**
         * P1 隔离加固：运行时探测 AssetManager.addAssetPath 私有 API 是否仍可反射调用。
         * 探测结果进程内缓存一次。未来 Android 封禁该 API 时，本次探测失败将走降级路径。
         */
        @JvmStatic
        fun probeAssetPathApi(): Boolean {
            addAssetPathAvailable?.let { return it }
            val available = try {
                AssetManager::class.java.getDeclaredMethod("addAssetPath", String::class.java) != null
            } catch (e: Throwable) {
                Log.e(TAG, "AssetManager.addAssetPath 私有 API 不可用: ${e.javaClass.simpleName}: ${e.message}")
                false
            }
            addAssetPathAvailable = available
            Log.i(TAG, "addAssetPath 可用性探测结果: $available")
            return available
        }
    }

    fun loadResources(moduleId: String, apkPath: String): ModuleResources? {
        resourceCache[moduleId]?.let { return it }

        if (!File(apkPath).exists()) {
            Log.e(TAG, "APK file not found: $apkPath")
            return null
        }

        // P1 降级路径：私有 API 被封禁时不再尝试加载外部资源，记录一次指标后返回 null；
        // 模块将以宿主资源 + 模块 ClassLoader 降级运行（无模块自带布局/主题）。
        if (!probeAssetPathApi()) {
            Log.w(TAG, "MODULE_RESOURCE_FALLBACK module=$moduleId api=addAssetPath（返回 null，模块降级为宿主资源）")
            return null
        }

        return try {
            // ===== 反射加载外部 APK 资源（私有 API） =====
            // 说明：
            // 1. AssetManager.addAssetPath 是 Android 私有 API（@hide 标注），未在公开 SDK 暴露
            // 2. 目前该 API 在所有 Android 版本（含 14/15）均可用，是加载外部 APK 资源的事实标准方案
            // 3. 未来 Android 版本可能通过非 SDK 接口黑名单（darkgreylist/maxtargetapi）封禁该 API，
            //    届时需切换备选方案
            // 4. 备选方案：使用 dalvik.system.PathClassLoader 加载 APK，再通过 PackageManager
            //    获取资源；缺点是效率较低（需解析整个 APK 包信息），且对未安装 APK 支持有限
            // 5. 反射失败时记录详细日志并返回 null，调用方应处理资源加载失败的情况
            @Suppress("DiscouragedPrivateApi")
            val assetManager: AssetManager = try {
                val addAssetPathMethod = AssetManager::class.java.getDeclaredMethod(
                    "addAssetPath", String::class.java
                )
                val am = AssetManager::class.java.getDeclaredConstructor().newInstance()
                // addAssetPath 返回 cookie（正整数）表示成功，返回 -1 或 0 表示路径无效/加载失败
                val result = addAssetPathMethod.invoke(am, apkPath) as? Int ?: -1
                if (result <= 0) {
                    Log.e(
                        TAG,
                        "addAssetPath returned invalid result: $result, apk: $apkPath, module: $moduleId"
                    )
                    return null
                }
                // 同时添加宿主 APK 路径，使模块 Resources 能访问宿主的 Material 主题等资源。
                // 否则模块布局 inflate 时，引用宿主主题属性（如 textAppearance）会触发
                // Resources$NotFoundException 闪退。
                val hostApkPath = context.applicationContext.applicationInfo.sourceDir
                if (hostApkPath != null) {
                    try {
                        addAssetPathMethod.invoke(am, hostApkPath)
                    } catch (e: Exception) {
                        Log.w(TAG, "Failed to add host APK path to module AssetManager: ${e.message}")
                    }
                }
                am
            } catch (e: NoSuchMethodException) {
                // 私有 API 在未来 Android 版本可能被封禁，此处捕获后返回 null 避免崩溃
                Log.e(
                    TAG,
                    "AssetManager.addAssetPath not found on this Android version, " +
                        "private API may have been blocked. Module: $moduleId", e
                )
                return null
            } catch (e: IllegalAccessException) {
                Log.e(
                    TAG,
                    "Illegal access to AssetManager.addAssetPath for module: $moduleId", e
                )
                return null
            } catch (e: java.lang.reflect.InvocationTargetException) {
                Log.e(
                    TAG,
                    "AssetManager.addAssetPath threw target exception for module: $moduleId, " +
                        "apk: $apkPath", e
                )
                return null
            } catch (e: InstantiationException) {
                Log.e(
                    TAG,
                    "Failed to instantiate AssetManager for module: $moduleId", e
                )
                return null
            }

            val hostResources = context.applicationContext.resources
            val displayMetrics = hostResources.displayMetrics
            val config = hostResources.configuration

            @Suppress("DEPRECATION")
            val moduleResources = Resources(assetManager, displayMetrics, config)

            val moduleRes = ModuleResources(
                moduleId = moduleId,
                resources = moduleResources,
                assetManager = assetManager,
                packageName = getModulePackageName(apkPath, moduleResources)
            )

            resourceCache[moduleId] = moduleRes
            Log.d(TAG, "Resources loaded for module: $moduleId, package: ${moduleRes.packageName}")
            moduleRes
        } catch (e: Exception) {
            // 兜底捕获，确保任何未预期异常不会导致应用崩溃
            Log.e(TAG, "Failed to load resources for module: $moduleId", e)
            null
        }
    }

    fun getResources(moduleId: String): ModuleResources? {
        return resourceCache[moduleId]
    }

    fun unloadResources(moduleId: String) {
        resourceCache.remove(moduleId)
        Log.d(TAG, "Resources unloaded for module: $moduleId")
    }

    private fun getModulePackageName(apkPath: String, resources: Resources): String {
        // 优先通过 PackageManager 从 APK 的 AndroidManifest.xml 直接读取包名，最可靠。
        val pmPackageName = try {
            val pm = context.applicationContext.packageManager
            val info = pm.getPackageArchiveInfo(apkPath, 0)
            info?.packageName
        } catch (e: Exception) {
            null
        }
        if (!pmPackageName.isNullOrEmpty()) {
            return pmPackageName
        }
        // 回退方案：通过 Resources.getIdentifier 反查 app_name 所属包名
        return try {
            val appResId = resources.getIdentifier("app_name", "string", "")
            if (appResId != 0) {
                resources.getResourcePackageName(appResId)
            } else {
                ""
            }
        } catch (e: Exception) {
            ""
        }
    }

    data class ModuleResources(
        val moduleId: String,
        val resources: Resources,
        val assetManager: AssetManager,
        val packageName: String
    ) {
        fun getString(resName: String): String? {
            return try {
                val resId = resources.getIdentifier(resName, "string", packageName)
                if (resId != 0) resources.getString(resId) else null
            } catch (e: Exception) {
                null
            }
        }

        fun getDrawable(resName: String): Drawable? {
            return try {
                val resId = resources.getIdentifier(resName, "drawable", packageName)
                if (resId != 0) resources.getDrawable(resId, null) else null
            } catch (e: Exception) {
                null
            }
        }

        fun getLayoutResId(resName: String): Int {
            return try {
                resources.getIdentifier(resName, "layout", packageName)
            } catch (e: Exception) {
                0
            }
        }

        fun getColor(resName: String): Int {
            return try {
                val resId = resources.getIdentifier(resName, "color", packageName)
                if (resId != 0) resources.getColor(resId, null) else 0
            } catch (e: Exception) {
                0
            }
        }

        fun getDimension(resName: String): Float {
            return try {
                val resId = resources.getIdentifier(resName, "dimen", packageName)
                if (resId != 0) resources.getDimension(resId) else 0f
            } catch (e: Exception) {
                0f
            }
        }

        fun getResId(resName: String, defType: String): Int {
            return try {
                resources.getIdentifier(resName, defType, packageName)
            } catch (e: Exception) {
                0
            }
        }
    }
}